### PR TITLE
task/JM-6700: Update 0008 - courtroom.sql

### PR DESCRIPTION
**Pull request by CCHAPMAN01**
Jira link (if applicable)

https://centralgovernmentcgi.atlassian.net/browse/JM-6700

### Change description ###
- update courtroom table migration script - fix constraint violations 

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines

Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by ChrisWCGI** [comment]
The new table has loc_code but the Heritage table has owner and this script is migrating the owner value to the loc-code value. 

Are we expecting the new table to have a list of court rooms under the loc_code for each and every court or just a list for each primary court? The data we are migrating from Heritage has one list of court rooms for each primary court and that list will include court rooms for any satellite courts. The only way of telling which courts rooms are for the primary vs which ones are for the satellite is by whatever free text the users have put in the description e,g, The Cardiff list has rooms that are prefixed with NPT for Newport.

If we want to split the list into separate lists for a primary and each of it's satellites we will need to ask HMCTS to identify which rooms belong to which courts for those courts that satellites.

**Comment by ChrisWCGI** [comment]
The LOCATION table for Derby (419) does not have a row for Asmbly nor does it have a row for ChkOut. All other courts have these entries. Asmbly is the default location specified in the COURT_LOCATION table for all courts. ChkOut is the location that is set on JUROR_POOL.LOCATION when a juror is checked out. Will this missing data for Derby be a problem?

**Comment by CCHAPMAN01** [comment]
> The LOCATION table for Derby (419) does not have a row for Asmbly nor does it have a row for ChkOut. All other courts have these entries. Asmbly is the default location specified in the COURT_LOCATION table for all courts. ChkOut is the location that is set on JUROR_POOL.LOCATION when a juror is checked out. Will this missing data for Derby be a problem?

Have checked for referential integrity - the location field is not linked to the courtroom table (it's free text) - it is also not being updated with ChkOut or Asmbly for checkout/check in journeys ... it also doesn't appear to be being read  anywhere to send back to the front end to display to users. Only written to when empanelled.

**Comment by CCHAPMAN01** [comment]
> The new table has loc_code but the Heritage table has owner and this script is migrating the owner value to the loc-code value.
> 
> Are we expecting the new table to have a list of court rooms under the loc_code for each and every court or just a list for each primary court? The data we are migrating from Heritage has one list of court rooms for each primary court and that list will include court rooms for any satellite courts. The only way of telling which courts rooms are for the primary vs which ones are for the satellite is by whatever free text the users have put in the description e,g, The Cardiff list has rooms that are prefixed with NPT for Newport.
> 
> If we want to split the list into separate lists for a primary and each of it's satellites we will need to ask HMCTS to identify which rooms belong to which courts for those courts that satellites.

Yes - spoken to Liam and agreed we want to have a list for each court locaiton. Action for @ChrisWCGI to create a list of data to send to the client to help split the list by individual loc_code

**Comment by CCHAPMAN01** [comment]
> The live database has an entry for the Bureau that we don't need ( owner = '400'). Shouldn't we be filtering this Bureau row out here?

Yes + question to ask Liam/Doug - should we also filter out any Asmbly or ChkOut rooms from the existing Heritage dataset (during migration) so they don't appear as courtrooms to select from when creating a trial?

**Comment by ChrisWCGI** [comment]
> > The live database has an entry for the Bureau that we don't need ( owner = '400'). Shouldn't we be filtering this Bureau row out here?
> 
> Yes + question to ask Liam/Doug - should we also filter out any Asmbly or ChkOut rooms from the existing Heritage dataset (during migration) so they don't appear as courtrooms to select from when creating a trial?

Navail is another non room location not associated with any trials and as such needs to be filtered out.

**Comment by ChrisWCGI** [comment]
> > The new table has loc_code but the Heritage table has owner and this script is migrating the owner value to the loc-code value.
> > Are we expecting the new table to have a list of court rooms under the loc_code for each and every court or just a list for each primary court? The data we are migrating from Heritage has one list of court rooms for each primary court and that list will include court rooms for any satellite courts. The only way of telling which courts rooms are for the primary vs which ones are for the satellite is by whatever free text the users have put in the description e,g, The Cardiff list has rooms that are prefixed with NPT for Newport.
> > If we want to split the list into separate lists for a primary and each of it's satellites we will need to ask HMCTS to identify which rooms belong to which courts for those courts that satellites.
> 
> Yes - spoken to Liam and agreed we want to have a list for each court locaiton. Action for @ChrisWCGI to create a list of data to send to the client to help split the list by individual loc_code

Note that it is possible that we may be asked to include a location for the primary and all of it's satellites i.e. 1 row from location might need to map to multiple rows in courtroom e.g. a generic Court 1

**Comment by ChrisWCGI** [comment]
How about migrate from location to courtroom after having migrated the trial table. The migrated trial table will tell us which court rooms are needed for which courts. We could check current data to see if there are any court rooms that have not been used according to the cross reference we will create from juror history for the trial migration.

**Comment by ChrisWCGI** [comment]
> > The LOCATION table for Derby (419) does not have a row for Asmbly nor does it have a row for ChkOut. All other courts have these entries. Asmbly is the default location specified in the COURT_LOCATION table for all courts. ChkOut is the location that is set on JUROR_POOL.LOCATION when a juror is checked out. Will this missing data for Derby be a problem?
> 
> Have checked for referential integrity - the location field is not linked to the courtroom table (it's free text) - it is also not being updated with ChkOut or Asmbly for checkout/check in journeys ... it also doesn't appear to be being read anywhere to send back to the front end to display to users. Only written to when empanelled.

Doug has said the court room list should only have court rooms in it i.e. Asmbly and Chkout should not be there.

Note that the court_location table has a foreign key from assembly_room to courtroom. Removing Asmbly from courtroo will require this foreign key to also be removed.

File: Postgres old schema to new schema migration/0008 - courtroom.sql [hunk 1]
```diff
@@ -18,27 +18,34 @@ values ('0008', 'juror', 'location', 'juror_mod', 'courtroom');
...
 
-with target
-as
-(
-	insert into juror_mod.courtroom("owner",room_number,description)
+
+with target as (
+	insert into juror_mod.courtroom(loc_code,room_number,description)
 	select distinct 
 			l."owner",
```
Comment (CCHAPMAN01) [inline Postgres old schema to new schema migration/0008 - courtroom.sql:43]: owner is being mapped to court location code so the approach will be:
where a primary court has satellite locations - duplicate courtrooms across the primary and all it's satellite (similar to holiday)
where a primary court has no satellite locations then the mapping is simple: owner = loc_code
there will be a reporting step after trial/panel migration where we can identify which courtrooms at which locations are not used and we can present the list to the client to clean up

Comment (ChrisWCGI) [reply Postgres old schema to new schema migration/0008 - courtroom.sql]: Agree. Is this PR awaiting an updated script or will the above change be done in a separate PR?

Comment (CCHAPMAN01) [reply Postgres old schema to new schema migration/0008 - courtroom.sql]: The script has been updated to migrate copies of courtrooms across all satellites (where applicable)

the reporting step will come as part of the trial/panel work in a separate PR

this one should be ready for review 🤞

**Pull request merged**